_posts/2016-05-24-SpringIO16-Spring-Rest-Docs.md

```diff
@@ -0,0 +1,339 @@
+                responseFields(
+                        fieldWithPath("username").description("The employee unique database identifier").type(String.class),
+                        fieldWithPath("_links").description("links to other resources")
+                ));
+{% endhighlight %}
+
+
+### Request fields
+This defines your fields you request from the client.
+Normally it will complain if you neglected a field but with v1.1 Spring REST Docs has support for Relaxed Snippets.
+Since I use constraints I made my own method `withPath, this will add an extra column constraint to the documentation.
+{% highlight java %} 
+ private static class ConstrainedFields {
+         private final ConstraintDescriptions constraintDescriptions;
+         ConstrainedFields(Class<?> input) {
+             this.constraintDescriptions = new ConstraintDescriptions(input);
+         }
+         private FieldDescriptor withPath(String path) {
+             return fieldWithPath(path).attributes(key("constraints").value(StringUtils
+                     .collectionToDelimitedString(this.constraintDescriptions
+                             .descriptionsForProperty(path), ". ")));
+         }
+     }
+
+ @Test
+ public void postEmployee() throws Exception{
+         Employee employee = employeeRepository.findByUsernameIgnoreCase("Nivek");
+         employee.setId(null);
+         employee.setUsername("Keloggs");
+         String string = objectWriter.writeValueAsString(employee);
+ 
+         ConstrainedFields fields = new ConstrainedFields(Employee.class);
+ 
+         this.document.snippets(
+                 requestFields(
+                         fields.withPath("username").description("The employee unique database identifier"),
+                         fields.withPath("firstName").description("The employee's first name"),
+                         fields.withPath("lastName").description("The employee's last name"),
+                         ));
+ 
+         mockMvc.perform(post("/employees").content(string).contentType(MediaTypes.HAL_JSON).header("Authorization", authToken)).andExpect(status().isCreated()).andReturn().getResponse().getHeader("Location");
+     }
+                        
+      
+{% endhighlight %}
+
+### Response fields
+This defines the result after consultation of a resource.
+{% highlight java %}
+  this.document.snippets(
+              responseFields(
+                       fieldWithPath("username").description("The employee unique database identifier"),
+                       fieldWithPath("firstName").description("The employee's first name"),
+                       fieldWithPath("lastName").description("The employee's last name"),
+                           ));
+{% endhighlight %}
+
+
+### Request/response headers
+Defines your request/response headers in your api. 
+This is useful when there are extra headers to set. 
+If the request has to involve an authorization header for security you can add this header to your document.
```

Comment:
When the request has to involve an authorization header for security reasons, you can add this header to your document.